### PR TITLE
MPI-4: MPI_Comm_idup_with_info

### PR DESCRIPTION
**Pull request by pavanbalaji**
## Pull Request Description

Initial implementation of `MPI_Comm_idup_with_info`, although the function is currently named with `MPIX_`.

Fixes https://github.com/pmodels/mpich/issues/4893

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by pavanbalaji** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by pavanbalaji** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by pavanbalaji** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by pavanbalaji** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

File: test/mpi/comm/idup_with_info.c [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    MPI_Wait(&reqs[0], MPI_STATUS_IGNORE);
+    errs += run_tests(newcomm1);
+
+    MPIX_Comm_idup_with_info(MPI_COMM_WORLD, MPI_INFO_NULL, &newcomm2, &reqs[0]);
+    MPI_Wait(&reqs[0], MPI_STATUS_IGNORE);
+    errs += run_tests(newcomm2);
+
+    /* Dup with info keys in mixed order */
+    MPI_Info_create(&info);
+    MPI_Info_set(info, (char *) "host", (char *) "myhost.myorg.org");
```
Comment (raffenet) [inline test/mpi/comm/idup_with_info.c:90]: should we check to see if these info are set on the new communicators?

File: test/mpi/comm/idup_with_info.c [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    MPI_Waitall(2, reqs, MPI_STATUSES_IGNORE);
+
+    errs += run_tests(newcomm3);
+    errs += run_tests(newcomm4);
+
+    MPI_Info_free(&info);
+    MPI_Comm_free(&newcomm4);
+    MPI_Comm_free(&newcomm3);
+
+    /* Dup with deleted info keys in mixed order */
```
Comment (raffenet) [inline test/mpi/comm/idup_with_info.c:110]: deleted info keys?

File: test/mpi/comm/idup_with_info.c (hunk 1 above)
Comment (pavanbalaji) [reply test/mpi/comm/idup_with_info.c:90]: It's hard to check unless we make the test mpich specific.  Other MPI implementations are not required to set them.  But I can add another test in the impls directory to check the info keys.

File: test/mpi/comm/idup_with_info.c (hunk 2 above)
Comment (pavanbalaji) [reply test/mpi/comm/idup_with_info.c:110]: The info keys are deleted before the communicator is created.

Comment (raffenet) [reply test/mpi/comm/idup_with_info.c:110]: Do you mean the info object is freed before the creation requests are waited on?

File: test/mpi/comm/idup_with_info.c (hunk 1 above)
Comment (raffenet) [reply test/mpi/comm/idup_with_info.c:90]: Ah, right. This is OK then.

File: test/mpi/comm/idup_with_info.c (hunk 2 above)
Comment (pavanbalaji) [reply test/mpi/comm/idup_with_info.c:110]: Right.  The `MPI_Waitall` is after the `MPI_Info_free`.

Comment (raffenet) [reply test/mpi/comm/idup_with_info.c:110]: Can you change the comment to make that clearer? "Delete" is a specific word that evokes `MPI_Info_delete`, which this test does not use.

**Comment by pavanbalaji** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Review by raffenet — APPROVED** [review]

**Pull request merged**